contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,269 @@
 pragma solidity >=0.8.13;
 
 import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
-
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
 import {ERC1155Singleton} from "./ERC1155Singleton.sol";
 import {IERC1155Singleton} from "./IERC1155Singleton.sol";
 import {IRegistry} from "./IRegistry.sol";
 import {IRegistryDatastore} from "./IRegistryDatastore.sol";
-import {IRegistryMetadata} from "./IRegistryMetadata.sol";
 import {BaseRegistry} from "./BaseRegistry.sol";
+import {EnhancedAccessControl} from "./EnhancedAccessControl.sol";
+import {MetadataMixin} from "./MetadataMixin.sol";
+import {IRegistryMetadata} from "./IRegistryMetadata.sol";
+import {SimpleRegistryMetadata} from "./SimpleRegistryMetadata.sol";
+import {NameUtils} from "../utils/NameUtils.sol";
+import {IPermissionedRegistry} from "./IPermissionedRegistry.sol";
+
+contract PermissionedRegistry is IPermissionedRegistry, BaseRegistry, EnhancedAccessControl, MetadataMixin {
+    event TokenRegenerated(uint256 oldTokenId, uint256 newTokenId);
+
+    mapping(uint256 => TokenObserver) public tokenObservers;
+
+    uint256 private constant ROLE_REGISTRAR = 1 << 0;
+    uint256 private constant ROLE_REGISTRAR_ADMIN = ROLE_REGISTRAR << 128;
 
-abstract contract PermissionedRegistry is BaseRegistry {
-    uint96 public constant FLAGS_MASK = 0x7;
-    uint96 public constant FLAG_SUBREGISTRY_LOCKED = 0x1;
-    uint96 public constant FLAG_RESOLVER_LOCKED = 0x2;
-    uint96 public constant FLAG_FLAGS_LOCKED = 0x4;
+    uint256 private constant ROLE_RENEW = 1 << 1;
+    uint256 private constant ROLE_RENEW_ADMIN = ROLE_RENEW << 128;
 
-    constructor(IRegistryDatastore _datastore) BaseRegistry(_datastore) {
+    uint256 private constant ROLE_SET_SUBREGISTRY = 1 << 2;
+    uint256 private constant ROLE_SET_SUBREGISTRY_ADMIN = ROLE_SET_SUBREGISTRY << 128;
+
+    uint256 private constant ROLE_SET_RESOLVER = 1 << 3;
+    uint256 private constant ROLE_SET_RESOLVER_ADMIN = ROLE_SET_RESOLVER << 128;
+
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER = 1 << 4;
+    uint256 private constant ROLE_SET_TOKEN_OBSERVER_ADMIN = ROLE_SET_TOKEN_OBSERVER << 128;
+
+    modifier onlyNonExpiredTokenRoles(uint256 tokenId, uint256 roleBitmap) {
+        _checkRoles(tokenIdResource(tokenId), roleBitmap, _msgSender());
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        if (expires < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        _;
     }
 
-    function _setFlags(uint256 tokenId, uint96 _flags)
-        internal
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    constructor(IRegistryDatastore _datastore, IRegistryMetadata _metadata) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, ALL_ROLES, _msgSender(), false);
```

Comment:
When instantiating this in a trustless context - for example, the .eth registry - we need to revoke or not set `ROLE_SET_SUBREGISTRY` and `ROLE_SET_RESOLVER` and `ROLE_SET_TOKEN_OBSERVER` for the contract owner. We could do this by calling the relevant method after deployment, but perhaps it would be cleaner to add a `deployerRoles` argument to the constructor?